### PR TITLE
added ravel/unravel operators

### LINKED ISSUE #10203: [Operator] unravel_index and ravel_multi_index
As requested by @fhieber, we should consider to support [unravel_index](https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.unravel_index.html) and it's inverse function [ravel_multi_index](https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.ravel_multi_index.html) in numpy. This feature is recently added to TF (https://github.com/tensorflow/tensorflow/pull/14895).

### PR DESCRIPTION
**Pull request by asmushetzel**
## Description ##
This resolves #10203 by implementing ravel_multi_index and unravel_index operators. The operators perform the same functionality as the corresponding operators in numpy.
This enables writing fully symbolic operations that involve ravel/unravel of indices. 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- [X] Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [ X] Code is well-documented: 
- [X] For user-facing API changes, API doc string has been updated. 
- [ X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

File: src/operator/tensor/ravel.h [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+  }
+};
+
+struct unravel_index {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i, index_t N, index_t ndim, index_t *shape,
+                                  DType *unravelled, DType *ravelled) {
+    index_t idx(ravelled[i]);
+    #pragma unroll
+    for (int j = ndim; j--; ) {
```
Comment (anirudh2290) [inline src/operator/tensor/ravel.h]: I find this style a little confusing and would prefer `for(int j = ndim; j >= 0; j--)` . What is the motivation ?

**Comment by fhieber** [comment]
Is this ready to merge? We have an important use case for these ops and would like to test them soon.

**Comment by asmushetzel** [comment]
Ready from my point of view.

**Comment by asmushetzel** [comment]
Changed input names from "A" to "data" as requested.

**Pull request merged**